internal/scm/azuredevops/azuredevops.go

```diff
@@ -0,0 +1,540 @@
+		return scm.PullRequestStatusClosed
+	case *pr.MergeStatus == git.PullRequestAsyncStatusValues.Conflicts ||
+		*pr.MergeStatus == git.PullRequestAsyncStatusValues.Failure ||
+		*pr.MergeStatus == git.PullRequestAsyncStatusValues.RejectedByPolicy:
+		return scm.PullRequestStatusError
+	case *pr.Status == git.PullRequestStatusValues.Active:
+		return scm.PullRequestStatusPending
+	default:
+		return scm.PullRequestStatusUnknown
+	}
+}
+
+// GetOpenPullRequest gets a pull request for one specific repository
+func (a *AzureDevOps) GetOpenPullRequest(ctx context.Context, repo scm.Repository, branchName string) (scm.PullRequest, error) {
+	r := repo.(repository)
+
+	prs, err := a.gitClient.GetPullRequests(ctx, git.GetPullRequestsArgs{
+		RepositoryId: &r.rid,
+		SearchCriteria: &git.GitPullRequestSearchCriteria{
+			Status:        &git.PullRequestStatusValues.Active,
+			SourceRefName: &branchName,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(*prs) == 0 {
+		return nil, nil
+	}
+
+	return convertPullRequest(&(*prs)[0]), nil
+}
+
+// MergePullRequest merges a pull request
+func (a *AzureDevOps) MergePullRequest(ctx context.Context, pullReq scm.PullRequest) error {
+	pr := pullReq.(pullRequest)
+
+	_, err := a.gitClient.UpdatePullRequest(ctx, git.UpdatePullRequestArgs{
+		GitPullRequestToUpdate: &git.GitPullRequest{
+			Status: &git.PullRequestStatusValues.Completed,
+			LastMergeSourceCommit: &git.GitCommitRef{
+				CommitId: &pr.lastMergeSourceCommitID,
+			},
+		},
+		RepositoryId:  &pr.repoID,
+		PullRequestId: &pr.id,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ClosePullRequest closes a pull request
+func (a *AzureDevOps) ClosePullRequest(ctx context.Context, pullReq scm.PullRequest) error {
+	pr := pullReq.(pullRequest)
+
+	_, err := a.gitClient.UpdatePullRequest(ctx, git.UpdatePullRequestArgs{
+		GitPullRequestToUpdate: &git.GitPullRequest{
+			Status: &git.PullRequestStatusValues.Abandoned,
+		},
+		RepositoryId:  &pr.repoID,
+		PullRequestId: &pr.id,
+	})
+	if err != nil {
+		return err
+	}
+
+	newObjectID := "0000000000000000000000000000000000000000"
+	refUpdate := git.GitRefUpdate{
+		Name:        &pr.branchName,
+		OldObjectId: &pr.lastMergeSourceCommitID,
+		NewObjectId: &newObjectID,
+	}
+
+	refUpdates := []git.GitRefUpdate{refUpdate}
+
+	_, err = a.gitClient.UpdateRefs(ctx, git.UpdateRefsArgs{
+		RepositoryId: &pr.repoID,
+		RefUpdates:   &refUpdates,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ForkRepository forks a project
+func (a *AzureDevOps) ForkRepository(ctx context.Context, repo scm.Repository, newProject string) (scm.Repository, error) {
+	r := repo.(repository)
+
+	currentUser, err := a.getCurrentUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	np := newProject
+	if np == "" {
+		np = r.ownerName
+	}
+
+	repoName := r.name + "." + strings.ReplaceAll(currentUser, " ", ".")
```

Comment:
Forks can actually only be made to Projects in Azure devops.  In the lines above, I am appending the username to the end of the repo name so that multiple forks of the repo can be forked into the same project.  The project to fork the repo into gets passed into this method through the newProject variable.

Reply:
Might this behavior be something users want to configure? I'm fine if you want to start with this, and potentially add it later.